### PR TITLE
Add read-only mode and query execution capabilities

### LINKED ISSUE #19: Full support for executing ad-hoc SQL queries is required
including INSERT, UPDATE, DELETE, and SELECT operations against the target database.

### PR DESCRIPTION
**Pull request by kevingatera**
This PR introduces read-only mode as a default security feature to prevent accidental write operations, adds the run_sql_query tool for executing SELECT statements, and implements a wrap_untrusted function to mitigate prompt injection risks. Changes update the DatabaseContext and DatabaseConnector classes to enforce read-only checks.

Please see: README documentation for configuration and usage.

This resolves #19 and is also a more flexible implementation of #17

### CODE REVIEW

File: db_context/database.py [hunk 1]
```diff
@@ -686,45 +703,112 @@ async def search_in_database(self, search_term: str, limit: int = 20) -> List[st
...
-                    "name": column_name,
-                    "type": data_type,
-                    "nullable": nullable == 'Y'
-                })
+            results = {}
+
+            # Optimized query to search for columns in the specified tables
+            # Using bind variables for table names is not directly supported in a single query like this,
+            # so we iterate. But we can optimize the query itself for each table.
+            for table_name in table_names:
```
Comment (danielmeppiel) [inline db_context/database.py:716]: I am just afraid this loop may let the LLM enter a pretty long list of tables and perform poorly. Any insights on that? Should this PR include this change in the scope?

File: db_context/database.py [hunk 2]
```diff
@@ -686,45 +703,112 @@ async def search_in_database(self, search_term: str, limit: int = 20) -> List[st
...
+            sql: The SQL query to execute.
+            params: A dictionary of bind parameters for the query.
+            max_rows: The maximum number of rows to return.
+            
+        Returns:
+            A dictionary containing the query results, including column definitions and rows.
+        """
+        conn = await self.get_connection()
+        try:
+            cursor = conn.cursor()
```
Comment (danielmeppiel) [inline db_context/database.py:756]: We shouldn't instantiate and manage cursors in functions, but use the helper functions _execute_cursor and _execute_cursor_no_fetch. They ensure we manage the thick vs thin modes. I think that's also the ideal place to implement read_only checks for security, as the single places where we touch cursors and ultimately send the query for execution.

File: db_context/database.py [hunk 3]
```diff
@@ -686,45 +703,112 @@ async def search_in_database(self, search_term: str, limit: int = 20) -> List[st
...
-    
+
     async def explain_query_plan(self, query: str) -> Dict[str, Any]:
-        """Get execution plan for a SQL query"""
+        """
+        Get the execution plan for a given SQL query and provide optimization suggestions.
+        This tool uses 'EXPLAIN PLAN FOR' to analyze the query without executing it.
+        """
+        if self.read_only:
+            return {
```
Comment (danielmeppiel) [inline db_context/database.py:806]: We should probably raise an error and handle on client side? Returning a formatted JSON object seems to be mixing concerns.

Comment (danielmeppiel) [reply db_context/database.py:806]: For consistency, use raise PermissionError("Read-only mode: write operations are disabled") by calling the assert_write_allowed. This should be the only right way to check for write permissions accross the database.py module.

File: db_context/database.py [hunk 4]
```diff
@@ -686,45 +703,112 @@ async def search_in_database(self, search_term: str, limit: int = 20) -> List[st
...
+        Returns:
+            A dictionary containing the query results, including column definitions and rows.
+        """
+        conn = await self.get_connection()
+        try:
+            cursor = conn.cursor()
             
-            return result
+            # Disallow DML/DDL in read-only mode
+            if self.read_only and not self._is_select_query(sql):
```
Comment (danielmeppiel) [inline db_context/database.py:759]: Maybe generalize assert_write_allowed with a function to check the executability of a query. It could both check the nature of the query (DDL/DML or just read) plus asserting if there are write permissions if it is a DDL/DML operation. This way we have a consistent unique way to check for "executability" across the file and simplify these kind of checks to a single function call which will also manage raising errors.

File: main.py [hunk 5]
```diff
@@ -607,5 +620,45 @@ async def get_related_tables(table_name: str, ctx: Context) -> str:
...
+        
+        # Header
+        formatted_result = "| " + " | ".join(headers) + " |\n"
+        # Separator
+        formatted_result += "| " + " | ".join(["---"] * len(headers)) + " |\n"
+        # Rows
+        for row in rows:
+            formatted_result += "| " + " | ".join(str(row[h]) for h in headers) + " |\n"
+            
+        return wrap_untrusted(formatted_result)
```
Comment (danielmeppiel) [inline main.py:658]: We seem to be using wrap_untrusted in only 2 functions. Is there a reason not to expand to all the other tools?

File: main.py [hunk 6]
```diff
@@ -607,5 +620,45 @@ async def get_related_tables(table_name: str, ctx: Context) -> str:
...
+        
+        if not result["rows"]:
+            return "Query executed successfully, but returned no rows."
+            
+        # Format the result as a markdown table
+        headers = result["columns"]
+        rows = result["rows"]
+        
+        # Header
+        formatted_result = "| " + " | ".join(headers) + " |\n"
```
Comment (danielmeppiel) [inline main.py:651]: Formatting logic should be externalized for consistency into db_context/schema/formatter.py

File: main.py [hunk 7]
```diff
@@ -607,5 +620,45 @@ async def get_related_tables(table_name: str, ctx: Context) -> str:
...
+        max_rows: The maximum number of rows to return (default is 100).
+
+    Returns:
+        A string containing the query results in a formatted table, or an error message if the query fails.
+    """
+    db_context: DatabaseContext = ctx.request_context.lifespan_context
+    try:
+        result = await db_context.run_sql_query(sql, max_rows=max_rows)
+        
+        if "error" in result:
```
Comment (danielmeppiel) [inline main.py:640]: This should rather be a try catch block where we catch the exception coming up from the database API.

File: db_context/database.py [hunk 8]
```diff
@@ -686,45 +703,112 @@ async def search_in_database(self, search_term: str, limit: int = 20) -> List[st
...
+        """
+        Get the execution plan for a given SQL query and provide optimization suggestions.
+        This tool uses 'EXPLAIN PLAN FOR' to analyze the query without executing it.
+        """
+        if self.read_only:
+            return {
+                "execution_plan": [],
+                "optimization_suggestions": [],
+                "error": "Explain plan is disabled in read-only mode."
+            }
```
[the remark this answers is not in the document]
Comment (kevingatera) [reply db_context/database.py:810]: True, but because it will create a temp table it would not really be a read only mode. Depends on what @danielmeppiel thinks

File: main.py (hunk 5 above)
Comment (kevingatera) [reply main.py:658]: I used it on only pieces where the data is coming from users. We don't expect the info about schema to contain user data that might contain prompt injections

File: main.py (hunk 6 above)
Comment (kevingatera) [reply main.py]: Change applied

File: db_context/database.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply db_context/database.py]: I agree with you @kevingatera

**Comment by kevingatera** [comment]
@danielmeppiel The PR is now ready. All concerns were addressed.

File: db_context/database.py [hunk 9]
```diff
@@ -103,18 +114,41 @@ async def _execute_cursor(self, cursor, sql: str, **params):
...
     async def _execute_cursor_no_fetch(self, cursor, sql: str, **params):
-        """Helper method for cursor operations that don't need fetching (e.g. DELETE, UPDATE)"""
+        """Helper method for statements that modify data (e.g. DELETE, UPDATE)."""
+        self._assert_query_executable(sql)
+        if self.thick_mode:
+            cursor.execute(sql, **params)
+        else:
+            await cursor.execute(sql, **params)
+    
+    async def _execute_cursor_with_fetch(self, cursor, sql: str, max_rows: int = 100, **params):
```
Comment (danielmeppiel) [inline db_context/database.py:135]: _execute_cursor_with_fetch and the older _execute_cursor are now a bit confusing. They seem to serve the same purpose, but yours accepts a max_rows parameter. 

I am thinking that we could merge both for consistency - and optionally allow the caller to define a max_rows value. Otherwise, fetchall instead of fetchmany. 

Either that or call this _execute_cursor_fetchmany and the legacy _execute_cursor renamed to _execute_cursor_fetchall. 

Whatever you prefer

File: db_context/database.py [hunk 10]
```diff
@@ -811,12 +905,48 @@ def _analyze_query_for_optimization(self, query: str) -> List[str]:
...
+           validation a full SQL parser should be used.
+        """
+        stripped = sql.lstrip()
+        upper_sql = stripped.upper()
+
+        if not (upper_sql.startswith("SELECT") or upper_sql.startswith("WITH")):
+            return False
+
+        # block attempts to chain multiple statements
+        if ";" in upper_sql:
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply db_context/database.py:926]: @kevingatera we should be parsing SQL properly to fully avoid SQL injection risk:

Here’s a succinct guide to implementing **sqlparse** in the project to improve SQL validation:

---

### How to Integrate `sqlparse` for Safer Query Validation

**1. Install the library (one-time):**
```bash
pip install sqlparse
```

**2. Import `sqlparse` where you validate SQL:**
```python
import sqlparse
```

**3. Use `sqlparse` to robustly check statement type:**
Replace the current `_is_select_query` heuristic with a parser-based approach:

```python
@staticmethod
def _is_select_query(sql: str) -> bool:
    """Return True if the statement is a single, pure SELECT/CTE."""
    statements = sqlparse.parse(sql)
    if len(statements) != 1:
        return False  # Prevent stacked queries
    stmt = statements[0]
    # Check if statement type is SELECT or WITH (for CTEs)
    first_token = stmt.token_first(skip_cm=True)
    return first_token and first_token.ttype in (sqlparse.tokens.DML,) and first_token.value.upper() in ("SELECT", "WITH")
```

**4. (Optional) Use `sqlparse` for further analysis:**
- You can detect if DML/DDL keywords exist outside string literals.
- You can prevent multi-statement queries more reliably.

**5. Review and test:**
- Ensure all SQL entrypoints use this method for query validation.
- Test with edge cases: semicolons inside literals, stacked statements, mixed case.

---

**Summary Table for Reviewer**

| Step | Action                    | Purpose                       | Code/Command               |
|------|---------------------------|-------------------------------|----------------------------|
| 1    | Install sqlparse          | Add dependency                | `pip install sqlparse`     |
| 2    | Import in code            | Enable parsing                | `import sqlparse`          |
| 3    | Replace heuristics        | Robust statement validation   | see code sample above       |
| 4    | Test & review             | Ensure correct query filtering| Unit tests, manual review   |

---

**Benefit:**  
This makes SELECT-only validation much more robust, catching stacked statements and ignoring semicolons inside literals, reducing SQL injection risk.

Let me know if you want ready-to-copy code for your specific file!

File: db_context/database.py [hunk 11]
```diff
@@ -811,12 +905,48 @@ def _analyze_query_for_optimization(self, query: str) -> List[str]:
...
+            return False
+
+        # simple keyword search for write operations
+        write_pattern = re.compile(r"\b(INSERT|UPDATE|DELETE|MERGE|CREATE|ALTER|DROP|TRUNCATE|GRANT|REVOKE)\b", re.IGNORECASE)
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply db_context/database.py:930]: @kevingatera again here we need to leverage `sqlparse`

File: db_context/database.py [hunk 12]
```diff
@@ -811,12 +905,48 @@ def _analyze_query_for_optimization(self, query: str) -> List[str]:
...
+        if write_pattern.search(upper_sql):
+            return False
+
+        return True
+
+    @staticmethod
+    def _is_write_operation(sql: str) -> bool:
```
Comment (danielmeppiel) [inline db_context/database.py:937]: @kevingatera let's use `sqlparse` for robustness and minimizing security risks:

```
import sqlparse

def is_write_operation(sql):
    write_ops = {'INSERT', 'UPDATE', 'DELETE', 'GRANT', 'CREATE', 'DROP', 'ALTER', 'TRUNCATE', 'REPLACE'}
    parsed = sqlparse.parse(sql)
    if not parsed:
        return False
    stmt = parsed[0]  # Only check the first statement
    for token in stmt.tokens:
        if token.ttype in sqlparse.tokens.Keyword.DML or token.ttype in sqlparse.tokens.Keyword.DDL:
            keyword = token.value.upper()
            return keyword in write_ops
        elif token.is_keyword:  # fallback for some keywords
            keyword = token.value.upper()
            return keyword in write_ops
    return False

# Example usage
print(is_write_operation("INSERT INTO table_name VALUES (1, 2, 3);"))  # True
print(is_write_operation("SELECT * FROM table_name;"))                # False
```

**Review by danielmeppiel — CHANGES_REQUESTED** [review]
Thanks for your latest changes @kevingatera !
Let's security-harden this with `sqlparse` library and we will be ready to merge

File: db_context/database.py [hunk 13]
```diff
@@ -811,12 +905,48 @@ def _analyze_query_for_optimization(self, query: str) -> List[str]:
...
+            "DELETE",
+            "MERGE",
+            "CREATE",
+            "ALTER",
+            "DROP",
+            "TRUNCATE",
+            "GRANT",
+            "REVOKE",
+        )
+        return any(stripped.startswith(keyword) for keyword in write_keywords)
```
[the remark this answers is not in the document]
Comment (kevingatera) [reply db_context/database.py]: No longer relevant

**Comment by kevingatera** [comment]
@danielmeppiel  The requested changes have been made. But, there are two items of note:
1. I ran `uv lock` and it included the new sqlparse but I think in the newer versions of uv it appends an upload time that resulted in a larger than expected line changes. From the cursory glance I made nothing important appears to have been disturbed other than bumping the revision, adding an unavoidable upload time per package, and adding sqlparse.
2.  The _assert_query_executable method only validates SQL when in read-only mode. When read-only mode is disabled, it doesn't check for stacked statements at all. I chose to keep it that way for non-read mode so the user/llm will have freedom of arbitrary commands.

**Comment by kevingatera** [comment]
@danielmeppiel is there anything else you'd like me to add/modify?
